custom_components/elasticsearch/es_doc_creator.py

```diff
@@ -157,33 +157,53 @@ def _state_to_entity_details(self, state: State) -> dict:
         """
         entity_details = self._entity_details.async_get(state.entity_id)
 
-        additions = {}
-
-        if entity_details is not None:
-            additions["device"] = {}
-
-            if entity_details.device:
-                additions["device"]["id"] = entity_details.device.id
-                additions["device"]["name"] = entity_details.device.name
+        if entity_details is None:
+            return {}
+
+        entity = entity_details.entity
+        entity_capabilities = entity.capabilities or {}
+        entity_area = entity_details.entity_area
+
+        entity_additions = {
+            "labels": entity_details.entity_label,
+            "name": entity.name or entity.original_name,
+            "friendly_name": state.name,
+            "platform": entity.platform,
+            "unit_of_measurement": str(entity.unit_of_measurement),
+            "area.id": entity_area.id if entity_area else None,
```

Comment:
I thought it was preferred to ingest the JSON object, instead of the flattened "dot notation" version. Is there a reason we need to change this?

Reply:
Elasticsearch dedots all objects by default. Here is the document in Elasticsearch:
```
                "_source": {
                    "@timestamp": "2024-05-16T22:38:25.722469Z",
                    "agent": {
                        "name": "My Home Assistant",
                        "type": "hass",
                        "version": "2024.4.0"
                    },
                    "data_stream": {
                        "dataset": "homeassistant.binary_sensor",
                        "namespace": "default",
                        "type": "metrics"
                    },
                    "ecs": {
                        "version": "1.0.0"
                    },
                    "event": {
                        "action": "State change",
                        "kind": "event",
                        "type": "change"
                    },
                    "hass": {
                        "entity": {
                            "attributes": {
                                "device_class": "battery",
                                "friendly_name": "friendly entity name"
                            },
                            "device": {
                                "area": {
                                    "id": "kitchen",
                                    "name": "Kitchen"
                                },
                                "class": "battery",
                                "floor": {
                                    "id": "floor_1",
                                    "name": "Floor 1"
                                },
                                "id": "cb17ffce4438675c4b10caba113732af",
                                "labels": "[]",
                                "name": "Kepler's observatory"
                            },
                            "domain": "binary_sensor",
                            "geo": {
                                "location": {
                                    "lat": 52.37313387449831,
                                    "lon": 4.890314619988203
                                }
                            },
                            "id": "binary_sensor.kepler_s_observatory_soil_monitor_battery_2",
                            "labels": "[]",
                            "name": "friendly entity name",
                            "platform": "ambient_station",
                            "value": "off",
                            "valueas": {
                                "boolean": false
                            }
                        },
                        "object_id": "kepler_s_observatory_soil_monitor_battery_2"
                    },
                    "host": {
                        "architecture": "aarch64",
                        "geo": {
                            "location": {
                                "lat": 52.37313387449831,
                                "lon": 4.890314619988203
                            }
                        },
                        "hostname": "9815f6fdaf32",
                        "os": {
                            "name": "Linux"
                        }
                    }
                }
            },
```

And we've technically been doing this for a long time: https://github.com/legrego/homeassistant-elasticsearch/blob/75151e1da2f0aee827c918d389db7853ad1a1319/custom_components/elastic/__init__.py#L421

```
        self._static_doc_properties = {
            'agent.name': config_dict['name'] if 'name' in config_dict else 'My Home Assistant',
            'agent.type': 'hass',
            'agent.version': system_info['version'],
            'ecs.version': '1.0.0',
            'host.geo.location': {
                'lat': config_dict['latitude'],
                'lon': config_dict['longitude']
            } if 'latitude' in config_dict else None,
            'host.architecture': system_info['arch'],
            'host.os.name': system_info['os_name'],
            'host.hostname': socket.gethostname(),
            'tags': config.get(CONF_TAGS)
        }
```

Reply:
I think the only downside is that if a user wants to use an ingest pipeline there is a chance they may have to dot expand via https://www.elastic.co/guide/en/elasticsearch/reference/current/dot-expand-processor.html to interact with the field using other processors.

Reply:
Ok, good enough for me.